scipy/sparse/csgraph/meson.build

```
@@ -1,27 +1,47 @@
-pyx_files = [
+pyx_files_for_c = [
   ['_flow', '_flow.pyx'],
   ['_matching', '_matching.pyx'],
   ['_min_spanning_tree', '_min_spanning_tree.pyx'],
   ['_reordering', '_reordering.pyx'],
-  ['_shortest_path', '_shortest_path.pyx'],
   ['_tools', '_tools.pyx'],
   ['_traversal', '_traversal.pyx']
 ]
+pyx_files_for_cpp = [
+  ['_shortest_path', '_shortest_path.pyx'],
+]
 
-cython_gen_csgraph = generator(cython,
+cython_gen_csgraph_for_c = generator(cython,
   arguments : cython_args,
   output : '@BASENAME@.c',
   depends : [_cython_tree, fs.copyfile('parameters.pxi')],
 )
 
-foreach pyx_file: pyx_files
+foreach pyx_file: pyx_files_for_c
   py3.extension_module(pyx_file[0],
-    cython_gen_csgraph.process(pyx_file[1]),
+    cython_gen_csgraph_for_c.process(pyx_file[1]),
     c_args: cython_c_args,
     dependencies: np_dep,
     link_args: version_link_args,
     install: true,
-    subdir: 'scipy/sparse/csgraph'
+    subdir: 'scipy/sparse/csgraph',
+  )
+endforeach
+
+cython_gen_csgraph_for_cpp = generator(cython,
+  arguments : cython_args,
```

This should fix it, thanks @da-woods !

```suggestion
  arguments : cython_cplus_args,
```

---

@rgommers in the root `meson.build` there is both

```meson
cython_args = ['-3', '--fast-fail', '--output-file', '@OUTPUT@', '--include-dir', '@BUILD_ROOT@', '@INPUT@']
cython_cplus_args = ['--cplus'] + cython_args
```
 and
```meson
cython_c_args += [use_math_defines]
cython_cpp_args = cython_c_args
```

are these supposed to be distinct? One was last touched by you, the other by @eli-schwartz . Perhaps `cython_cpp_args` is missing `--cplus`?

---

cython_cpp_args is definitely not the same thing, since those are the arguments passed to GCC, not to the cython transpiler.

---

In ab3a8bc9519ce1658351f6d04e44b9a38eb269ea

```diff
+cython_gen_csgraph = generator(cython,
+  arguments : cython_args,
+  output : '@BASENAME@.c',
+  depends : [_cython_tree, fs.copyfile('parameters.pxi')],
+)
+
 foreach pyx_file: pyx_files
   py3.extension_module(pyx_file[0],
-    cython_gen.process(pyx_file[1]),
+    cython_gen_csgraph.process(pyx_file[1]),
     c_args: cython_c_args,
```

So before it was using C mode, and now it is still using C mode, but you want to pass --cplus for this scenario...

If generating C++ code is the goal then this should be emitting files named `output : '@BASENAME@.cpp',` or it is going to break the compiler handling.

---

> If generating C++ code is the goal then this should be emitting files named output : '@BASENAME@.cpp'

I think that is done correctly in this PR, creating a new `cython_gen_csgraph_for_cpp`, separate from `cython_gen_csgraph`, which is renamed to `cython_gen_csgraph_for_c`. I think we just need to change to my original suggestion, of passing `cython_cplus_args` (which includes `--cplus`) here?

---

Ah yes, that makes sense. Indeed, any time you emit .cpp files with cython you should also pass the cython_cplus_args.